### PR TITLE
feat(nextra): support symlinked pages

### PR DESCRIPTION
**Pull request by lttb**
hey, thank you a lot for your work!

what do you think about supporting symlinks?

I believe it might be quite useful, especially for monorepos, where documentation might be distributed across the packages and not centralised under the pages of the documentation website.

say we have a monorepo with the following structure:

```
monorepo
├── docs
│   ├── pages
│   │   ├── index.md
│   │   ├── quick-start.md
│   │   ├── changelog.md -> ../CHANGELOG.md
│   │   └── components -> ../../components
│   └── CHANGELOG.md
├── components
│   ├── Button
│   │   ├── index.tsx
│   │   ├── Button.tsx
│   │   ├── Button.md
│   │   └── CHANGELOG.md
│   └── Dropdown
│       ├── index.tsx
│       ├── Dropdown.tsx
│       ├── Dropdown.md
│       └── CHANGELOG.md
└── package.json
```

and we want to display the documentation for the components, which are outside of the docs folder.

even though it's already possible, for example, to manually import `docs/CHANGELOG.md` from `docs/pages/changelog.md` to render the document outside of the pages folder, for components, it wouldn't be that simple to do so without simlink support, as components have a nested structure, and it'll require quite a lot of manual mapping.

I've reproduced a simple example with a similar structure and proposed implementation here: https://github.com/lttb/nextra/compare/symlink-support...lttb:nextra:symlink-support-example?expand=1

I apologize for the lack of tests. I'm happy to add them if you think this feature makes sense (although I'm not totally sure what would be the best way to organise them at the moment)

thanks!

### CODE REVIEW

**Review by B2o5T — COMMENTED** [review]
I think this feature make sense, I already had errors from Nextra in one project where some pages were symlinked

Does there is a simple way to determine from `context` a real path of file instead of retrieving from `context.resourcePath`?

**Comment by lttb** [comment]
@B2o5T I've pushed a commit (6398cec) with another approach based on webpack module's [resourceResolveData.relativePath](https://github.com/webpack/webpack/blob/3c0af27ec6d5dc0ad33243a018068e5f1ab559cd/types.d.ts#L7685), to retrieve a symlink target path.

it looks a bit hacky, but I didn't find a simpler way to retrieve a symlink's target path just in the loader context yet.

although this implementation makes fewer structural changes, I feel like the original approach was a bit more straightforward and might be less error-prone.

what do you think?

**Comment by B2o5T** [comment]
Looks good can you add some tests with symlinks for `collectFiles` function? 🙏

**Comment by lttb** [comment]
hey @B2o5T, I've added some tests:
- added `swr-site/pages/about/changelog` as a symlink (used to generate a snapshot for `context.test.ts` pages collection)
- added `nextra/__test__/fixture/page-maps/folder-with-symlinks` and updated `page-map.test.ts` with a test to check the collected symlinked pages

hope it works!

File: packages/nextra/src/plugin.ts [hunk 1]
```diff
@@ -57,13 +59,25 @@ export async function collectFiles(
   dir: string,
   locales = DEFAULT_LOCALES,
   route = '/',
-  fileMap: FileMap = Object.create(null)
+  fileMap: FileMap = Object.create(null),
+  isFollowingSymlink = false
```
Comment (B2o5T) [inline packages/nextra/src/plugin.ts:63]: this function now accepts too many arguments, should be refactored to object in the future :)

Comment (lttb) [reply packages/nextra/src/plugin.ts:63]: agree, I wasn't sure about including bigger refactoring in the scope of this work. happy to open a following pr with a refactored signature

**Comment by lttb** [comment]
@B2o5T it seems that tests for `loader.test.ts` are failing because `examples/swr-site` should be built first - I think I had a similar issue locally

**Review by B2o5T — APPROVED** [review]
I think is fine for the moment to include your feature in the next release. I would be appreciated if you take a look at mentioned problems with symlinks outside cwd.

Also, I was surprised by your work, I liked it a lot! Great job

I'll make CI pass and will merge it!

Thank you @lttb!

File: examples/swr-site/pages/about/changelog.en-US.md [hunk 2]
```diff
@@ -0,0 +1 @@
+../../../../packages/nextra/CHANGELOG.md
```
[the remark this answers is not in the document]
Comment (lttb) [reply examples/swr-site/pages/about/changelog.en-US.md]: @B2o5T thank you for investigating! interestingly, at least loader tests pass with top-level README (which is still outside of `swc-site` cwd), for example,

```diff
- ../../../../packages/nextra/CHANGELOG.md
+ ../../../../README.md
```

image

although hydration failed too 

image

so yeah, looks like it's related to the cwd somehow. I'll try to dive more to see if we can fix it.

and thank you very much for helping and releasing it!

[the remark this answers is not in the document]
Comment (B2o5T) [reply examples/swr-site/pages/about/changelog.en-US.md]: You are welcome!

**Pull request merged**